### DISCUSSION TITLE
How to get the NSWindow handle of the MainWindow on MacOS

### DISCUSSION THREAD

**Discussion by cbrunin** [AvaloniaUI/Avalonia#9757] [Q&A]
Hi,
I'm trying to get the window handle of the main window with the following code, but there is a cast issue.

          if (Application.Current.ApplicationLifetime is IClassicDesktopStyleApplicationLifetime mainWindow)
            {
                var toplevel = (TopLevel)mainWindow.MainWindow.GetVisualRoot();
                var win = ((IMacOSTopLevelPlatformHandle) toplevel.PlatformImpl).NSWindow;
             }

PlatformImpl cannot be casted to IMacOSTopLevelPlatformHandle. 

The application is running on Mac OS.

Any clue?
Thanks,

**Comment by maxkatz6** [comment]
> PlatformImpl cannot be casted to IMacOSTopLevelPlatformHandle.

It should be `(IMacOSTopLevelPlatformHandle)toplevel.PlatformImpl.Handle` and not `(IMacOSTopLevelPlatformHandle) toplevel.PlatformImpl`.

Also:
1. This code will fail anyway on non-mac targets. Better to use c# conditions like `if (Handle is IMacOSTopLevelPlatformHandle macHandle)`.
2. For simple scenarios it should be enough to just get IntPtr handle `toplevel.PlatformImpl.Handle.Handle` (IMacOSTopLevelPlatformHandle interface has more properties, but you might not need them).
